pmlc/dialect/comp/ir/ops.td

```diff
@@ -86,50 +99,59 @@ def COMP_DestroyExecEnv : COMP_Op<"destroy_execenv",
     Does not guarantee that submitted tasks finished executing.
   }];
 
-  let arguments = (ins COMP_AnyExecEnv:$execEnv);
+  let arguments = (ins Arg<COMP_AnyExecEnv, "", [MemFree<COMP_ExecEnvResource>]>:$execEnv);
 
   let assemblyFormat = [{
      $execEnv attr-dict `:` type($execEnv)
   }];
 }
 
-def COMP_ScheduleEnd : COMP_Op<"schedule_end", [Terminator]> {
-  let summary = "End of scheduling block.";
+def COMP_CreateKernel : COMP_Op<"create_kernel",
+                                [DeclareOpInterfaceMethods<COMP_ExecEnvOpInterface>]>
+                  , Arguments<(ins Arg<COMP_AnyExecEnv, "", [MemRead<COMP_ExecEnvResource>]>:$execEnv,
+                                   SymbolRefAttr:$kernelFunc)>
+                  , Results<(outs Res<COMP_Kernel, "", [MemAlloc<COMP_KernelResource>]>:$kernel)> {
+  let summary = "Creates an executable kernel.";
+  let assemblyFormat = [{
+    `on` $execEnv attr-dict `:` functional-type(operands, results)
+  }];
 }
 
-def COMP_ScheduleSingleBlock : SingleBlockImplicitTerminator<"ScheduleEnd">;
+def COMP_DestroyKernel : COMP_Op<"destroy_kernel",
+                                 [DeclareOpInterfaceMethods<COMP_ExecEnvOpInterface>]>
+                  , Arguments<(ins Arg<COMP_AnyExecEnv, "", [MemRead<COMP_ExecEnvResource>]>:$execEnv,
+                                   Arg<COMP_Kernel, "", [MemFree<COMP_KernelResource>]>:$kernel)> {
+  let summary = "Destroys an executable kernel.";
+  let assemblyFormat = [{
+    $kernel `on` $execEnv attr-dict `:` functional-type(operands, results)
+  }];
+}
 
-def COMP_ScheduleFunc : COMP_Op<"schedule_func",
-                                [COMP_ScheduleSingleBlock,
-                                 DeclareOpInterfaceMethods<COMP_ExecEnvOpInterface>,
-                                 DeclareOpInterfaceMethods<COMP_ScheduleOpInterface>]>
-                  , Arguments<(ins COMP_AnyExecEnv:$execEnv,
+def COMP_ScheduleCompute : COMP_Op<"schedule_compute",
```

Comment:
Maybe `schedule_kernel` just so its consistent with `alloc_kernel`, `dealloc_kernel`.

Reply:
IMHO -- this operation doesn't change the state of the kernel at all; it's not something one does to the kernel, and you can do it multiple times with the same kernel with various inputs, producing different completion events.  It's more like you're enqueuing some compute task to be run, and the kernel is an input to that compute task.

Reply:
We could call it `launch_kernel`?

Reply:
I agree with Rob on this one.  You might schedule computations with different grid sizes, buffers, etc, so it's not really one-to-one with a kernel, although each compute does have a unique kernel, so I think schedule_compute is a better name.

Reply:
Also other things that make events all use schedule_....